gax-grpc/src/main/java/com/google/api/gax/grpc/GrpcCallContext.java

```diff
@@ -170,12 +183,34 @@ public GrpcCallContext withStreamIdleTimeout(@Nullable Duration streamIdleTimeou
     }
 
     return new GrpcCallContext(
-        channel, callOptions, streamWaitTimeout, streamIdleTimeout, channelAffinity);
+        channel, callOptions, streamWaitTimeout, streamIdleTimeout, channelAffinity, extraHeaders);
   }
 
+  @BetaApi("The surface for channel affinity is not stable yet and may change in the future.")
   public GrpcCallContext withChannelAffinity(@Nullable Integer affinity) {
     return new GrpcCallContext(
-        channel, callOptions, streamWaitTimeout, streamIdleTimeout, affinity);
+        channel, callOptions, streamWaitTimeout, streamIdleTimeout, affinity, extraHeaders);
+  }
+
+  @BetaApi("The surface for extra headers is not stable yet and may change in the future.")
+  @Override
+  public GrpcCallContext withExtraHeaders(@Nullable Map<String, List<String>> extraHeaders) {
+    ImmutableListMultimap.Builder<String, String> newExtraHeadersBuilder =
```

Comment:
(sorry if this was already discussed before with other reviewers)
It looks like internally the headers are stored in `ImmutableListMultimap`, but on the surface it is `Map<String, List<String>>`. This causes unnecessary conversions in both directions here (in setter) and in `getExtraHeaders` (getter). Please strongly consider keeping internal representation and external same (i.e. just `Map<String, List<String>>`). It is not as fancy as `ImmutableListMultimap`, which is a right data structure here, but since we cannot use guava on the surface, to keep things simple, I think it is better to avoid unnecessary covnersions between equivalent data structures (`ImmutableListMap` and `Map<?, List<?>>`). Please check next comment (line 205) as an example of potential complications, when internal and external representaitons are different.

Reply:
That sounds very reasonable to me. @pongad Michael, do you have any input here?